### ISSUE TITLE
KAGGLE / Computer Vision Tutorials / Create image classifiers with TensorFlow and Keras, and explore convolutional neural networks.

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/Data-Analysis#2]
- Use modern deep-learning networks to build an image classifier with Keras
- Design your own custom convnet with reusable blocks
- Learn the fundamental ideas behind visual feature extraction
- Master the art of transfer learning to boost your models
- Utilize data augmentation to extend your dataset

### [**The Convolutional Classifier**]
A convnet used for image classification consists of two parts: a **convolutional base** and a **dense head**.
![image](https://user-images.githubusercontent.com/70500841/97659694-f90fdf80-1ab2-11eb-94ec-117f8cf40d78.png)

The parts of a convnet: image, base, head, class; input, extract, classify, output.
The base is used to **extract the features** from an image. It is formed primarily of layers performing the convolution operation, but often includes other kinds of layers as well. (You'll learn about these in the next lesson.)

The head is used to **determine the class** of the image. It is formed primarily of dense layers, but might include other layers like dropout.

What do we mean by visual feature? A feature could be a line, a color, a texture, a shape, a pattern -- or some complicated combination.

The whole process goes something like this:
![image](https://user-images.githubusercontent.com/70500841/97659748-1cd32580-1ab3-11eb-9af3-55460c627341.png)

### [**Training the Classifier**]
The goal of the network during training is to learn two things:

1. which features to extract from an image (base),
2. which class goes with what features (head).
These days, convnets are rarely trained from scratch. More often, we **reuse the base of a pretrained model**. To the pretrained base we then **attach an untrained head**. In other words, we reuse the part of a network that has already learned to do 1. Extract features, and attach to it some fresh layers to learn 2. Classify.

Attaching a new head to a trained base.
Because the head usually consists of only a few dense layers, very accurate classifiers can be created from relatively little data.

Reusing a pretrained model is a technique known as **transfer learning**. It is so effective, that almost every image classifier these days will make use of it.

### [**Feature Extraction**]
Before we get into the details of convolution, let's discuss the purpose of these layers in the network. We're going to see how these three operations (convolution, ReLU, and maximum pooling) are used to implement the feature extraction process.

The **feature extraction** performed by the base consists of **three basic operations**:

1. **Filter** an image for a particular feature (convolution)
2. **Detect** that feature within the filtered image (ReLU)
3. **Condense** the image to enhance the features (maximum pooling)

### [**Filter with Convolution**]

```
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Conv2D(filters=64, kernel_size=3), # activation is None
    # More layers follow
])
```

* filters parameters
-Weights (=Kernels)
-Activations (=Feature maps)

### [**Detect with ReLU**]
-rectifier function (=ReLU activation, ReLU function)

```
model = keras.Sequential([
    layers.Conv2D(filters=64, kernel_size=3, activation='relu')
    # More layers follow
])
```

### [**Condense with Maximum Pooling**]
```
 from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Conv2D(filters=64, kernel_size=3), # activation is None
    layers.MaxPool2D(pool_size=2),
    # More layers follow
])
```

```
import tensorflow as tf

image_condense = tf.nn.pool(
    input=image_detect, # image in the Detect step above
    window_shape=(2, 2),
    pooling_type='MAX',
    # we'll see what these do in the next lesson!
    strides=(2, 2),
    padding='SAME',
)

plt.figure(figsize=(6, 6))
plt.imshow(tf.squeeze(image_condense))
plt.axis('off')
plt.show();
```

### [**Translation Invariance**(불변성)] 
-> 함수의 입력이 바뀌어도 출력은 그대로 유지되어 바뀌지 않음. (ex. 강아지 사진에서 강아지가 어느 위치에 있건 상관없이 '강아지'를 출력함을 의미)

-maximum pooling creates **translation invariance** over **small distances**. This means that we would expect small shifts to disappear after repeated maximum pooling. If you run the cell multiple times, you can see the resulting image is always the same; the pooling operation destroys those small translations.
zero-pixel은 중요하지 않다고 배웠으나, 사실은 위치정보를 나타냄
가까운 거리의 점들은 polling하면 위치 정보가 손실될 수 있음. (??)

-max-pooing은 대표적인 small translation invariance 함수이다. 여러 픽셀 중 최대값을 가진 픽셀 하나를 출력하기 때문에 서로 다른 [1,0,0], [0,0,1] 두 입력을 넣어도 동일한 1을 출력하게 된다. 그래서 입 모양이 약간씩 다르거나 위치가 조금씩 다른 경우에도 동일한 입으로 인식하게 되는 것이다.

-여기서 중요한건 CNN모델이 convolution연산의 equivariance 특성에 파라미터 공유를 더하게 되면 나오는 결과이다. convolution + 파라미터 공유를 하면 equivariance를 통해 반대 특성인 translation invariance 특성 또한 갖게 된다.
파라미터를 공유하지 않으면 위치마다 다른 필터를 사용하기 때문에 하단에 있는 토끼 이미지만 학습했다면 상단의 있는 토끼 이미지의 경우 처음보는 특성으로 판단하여 토끼라고 판단하지 않을 수 있다. 즉 translation invariance특성을 갖지 못하게 된다는 것이다.

-translation invariance가 싫어요
translation invariance를 피하고 싶은 경우도 있을까?
물론이다. 각 특징들의 위치 정보가 중요한 경우 translation invariance를 막는 CNN 아키텍쳐들도 있다.
예를 들어 정면으로 정중앙에 얼굴만 있는 사진을 학습할 때 눈이 아래에 있으면 얼굴로 판단하면 안된다.
이러한 경우를 위해 Locally connected layers(unshared convolution)는 일반적인 Convolution 연산과 동일하지만 파라미터 공유를 하지 않는다. 즉 각 위치마다 다른 필터를 사용하기 때문에 동일한 특징이라도 위치에 따라 다른 필터를 학습을 하게 된다.
결국 필터가 위치 정보를 포함한다고 볼 수 있으며, 결과적으로 translation invariance를 버리게 된다.(max-pooling을 사용한다면 small translation invariance 특성은 가질 수 있다.)
또 다른 유명한 모델로 캡슐 네트워크가 있다. 캡슐 네트워크는 아예 새로운 방식으로 translation invariance문제를 해결한다.
아래 그림은 캡슐 네트워크 포스팅에서 가장 많이 나오는 그림으로 CNN의 translation invariance 단점을 꼬집은 경우다. 오른쪽 그림은 얼굴이라고 볼 수 없는 사진이지만 위치에 상관없이 얼굴의 부분적인 특징들이 모두 있기 때문에 얼굴이라고 잘못 판단한다는 예제이다.

* **Global Average Pooling**
GlobalAvgPool2D layer is often used as an alternative to some or all of the hidden Dense layers in the head of the network, like so:

```
model = keras.Sequential([
    pretrained_base,
    layers.GlobalAvgPool2D(),
    layers.Dense(1, activation='sigmoid'),
])
```

What is this layer doing? Notice that we no longer have the Flatten layer that usually comes after the base to transform the 2D feature data to 1D data needed by the classifier. Now the GlobalAvgPool2D layer is serving this function. But, instead of "unstacking" the feature (like Flatten), it simply replaces the entire feature map with its average value. Though very destructive, it often works quite well and has the advantage of reducing the number of parameters in the model.

### [**The Sliding Window**]
The convolution and pooling operations share a common feature: they are both performed over a **sliding window**. With convolution, this "window" is given by the dimensions of the kernel, the parameter **kernel_size**. With pooling, it is the pooling window, given by **pool_size**.

![image](https://user-images.githubusercontent.com/70500841/97666965-1e581a00-1ac2-11eb-8ebb-7ac6d6bdb487.png)
There are two additional parameters affecting both convolution and pooling layers -- these are the **strides** of the window and whether to use **padding** at the image edges. The **strides** parameter says how far the window should move at each step, and the **padding** parameter describes how we handle the pixels at the edges of the input.

With these two parameters, defining the two layers becomes:

```
from tensorflow import keras
from tensorflow.keras import layers

model = keras.Sequential([
    layers.Conv2D(filters=64,
                  kernel_size=3,
                  strides=1,
                  padding='same',
                  activation='relu'),
    layers.MaxPool2D(pool_size=2,
                     strides=1,
                     padding='same')
    # More layers follow
])
```

### [**Stride**] 
-> A convolution with strides of 3 is too coarse to produce a good feature map from it. (stride 값이 커질수록 이미지가 거칠어진다)
The distance the window moves at each step is called the **stride**. We need to specify the stride in both dimensions of the image: one for moving left to right and one for moving top to bottom. This animation shows **strides=(2, 2)**, a movement of 2 pixels each step.
![image](https://user-images.githubusercontent.com/70500841/97667116-7f7fed80-1ac2-11eb-9030-ecb15c18ab37.png)

What effect does the stride have? Whenever the stride in either direction is greater than 1, the sliding window will skip over some of the pixels in the input at each step.

### [**Padding**]
What the convolution does with these boundary values is determined by its padding parameter. In TensorFlow, you have two choices: either **padding='same'** or **padding='valid'.** There are trade-offs with each.
